File: net/Plane.cpp

using namespace Engine;

Shapes.Plane.Plane(Vector normal, double d, IMaterial material) :
	normal(normal),
	d(d),
	A( (normal.x != 0 ? -d / normal.x : 0                                              ),
	   (normal.x != 0 ? 0             : (normal.y != 0 ? -d / normal.y : 0            )),
	   (normal.x != 0 ? 0             : (normal.y != 0 ? 0             : -d / normal.z))
	   ),
	material(material)
{
}

Shapes.Plane.Plane(Vector a, Vector b, Vector A, IMaterial material) :
	normal(a.CrossProduct(b).Normalized),
	A(A),
	material(material)
{
	d = A.DotProduct(normal);
}

Shapes.Plane.Plane(double a, double b, double c, double d, IMaterial material):
	normal(new Vector(a, b, c).Normalized),
	d(d),
	material(material),
	A( (a != 0 ? -d / a : 0                         ),
	   (a != 0 ? 0      : (b != 0 ? -d / b : 0     )),
	   (a != 0 ? 0      : (b != 0 ? 0      : -d / c))
	   )
{
}

HitPoint Shapes.Plane.Intersection(Vector start, Vector direction)
{
	double divident = normal.DotProduct(direction);

	if(divident == 0)
	{
		return NULL;
	}

	double t = (start.DotProduct(normal) + d) / normal.DotProduct(direction);

	if(t < 0)
	{
		return NULL;
	}

	return new HitPoint(t, normal, material);
}

File: net/Shapes/Plane.cs
using Engine;

namespace Shapes;

public class Plane : IShape
{
	public readonly double d;
	public readonly Vector normal;
	public readonly Vector A;
	public readonly AABB aabb;
	public readonly bool oneSide;
	public readonly bool isSolid;

	public Plane(Vector normal, double d, bool oneSide = false, bool isSolid = false)
	{
		this.normal = normal;
		this.d = d;
		this.A = normal.x != 0 ? new Vector(-d / normal.x, 0, 0) :
				normal.y != 0 ? new Vector(0, -d / normal.y, 0) :
				new Vector(0, 0, -d / normal.z);
		this.oneSide = oneSide;
		this.isSolid = isSolid;
		aabb = AABB.MaxValue;
	}

	public Plane(Vector a, Vector b, Vector A, bool oneSide = false, bool isSolid = false)
	{
		this.normal = a.CrossProduct(b).Normalized;
		this.A = A;
		this.d = A.DotProduct(normal);
		this.oneSide = oneSide;
		this.isSolid = isSolid;
		aabb = AABB.MaxValue;
	}

	public Plane(double a, double b, double c, double d, bool oneSide = false, bool isSolid = false)
	{
		this.normal = new Vector(a, b, c).Normalized;
		this.d = d;
		this.A = a != 0
			? new Vector(-d / a, 0, 0)
			: (
				b != 0
				? new Vector(0, -d / b, 0)
				: new Vector(0, 0, -d / c)
			);
		this.oneSide = oneSide;
		this.isSolid = isSolid;
		aabb = AABB.MaxValue;
	}

	public ref readonly AABB AABB => ref aabb;

	public ShapeHitPoint Intersection(in Ray ray)
	{
		var divident = normal.DotProduct(ray.direction);

		if(divident == 0 || oneSide && divident < 0)
		{
			return null;
		}

		var t = (ray.start.DotProduct(normal) + d) / divident;

		if(t < 0)
		{
			return null;
		}

		return new ShapeHitPoint(ray, t, isSolid ? normal : (normal * Math.Sign(divident)));
	}
}
